vm_trace.c

```diff
@@ -1755,72 +1734,53 @@ rb_postponed_job_register_one(unsigned int flags, rb_postponed_job_func_t func,
 int
 rb_workqueue_register(unsigned flags, rb_postponed_job_func_t func, void *data)
 {
-    struct rb_workqueue_job *wq_job = malloc(sizeof(*wq_job));
-    rb_vm_t *vm = GET_VM();
-
-    if (!wq_job) return FALSE;
-    wq_job->job.func = func;
-    wq_job->job.data = data;
-
-    rb_nativethread_lock_lock(&vm->workqueue_lock);
-    ccan_list_add_tail(&vm->workqueue, &wq_job->jnode);
-    rb_nativethread_lock_unlock(&vm->workqueue_lock);
-
-    // TODO: current implementation affects only main ractor
-    RUBY_VM_SET_POSTPONED_JOB_INTERRUPT(rb_vm_main_ractor_ec(vm));
-
-    return TRUE;
+    return rb_postponed_job_register_impl(flags, func, data, 0);
 }
 
 void
 rb_postponed_job_flush(rb_vm_t *vm)
 {
+    size_t postponed_job_count, index;
+    rb_postponed_job_t *jobs;
+
+    rb_native_async_safe_mutex_lock(&vm->postponed_job_buffer->lock);
+    size_t job_bytes = sizeof(rb_postponed_job_t) * vm->postponed_job_buffer->job_count;
+    /* Important to use actual malloc, not ruby_xmalloc; we don't want this to throw any
+       ruby exceptions whilst we're holding the postponed buffer lock & have signals masked. */
+    jobs = malloc(job_bytes);
+    if (jobs) {
+        memcpy(jobs, vm->postponed_job_buffer->jobs, job_bytes);
+        postponed_job_count = vm->postponed_job_buffer->job_count;
+        vm->postponed_job_buffer->job_count = 0;
+    }
+    rb_native_async_safe_mutex_unlock(&vm->postponed_job_buffer->lock);
```

Comment:
I think you missed `free`ing the jobs after going through them :)

Reply:
Oops! great catch, thank you!